### PR TITLE
test: improve `TestConcurrentReadAndWrite` to verify read/write linerizability

### PR DESCRIPTION
**Pull request by ahrtr**
Validate the linearizablity: each reading transaction should read the same value written by previous writing transaction on the same key.

### CODE REVIEW

**Comment by ahrtr** [comment]
cc @ptabor @serathius

File: concurrent_test.go [hunk 1]
```diff
@@ -116,24 +119,19 @@ func concurrentReadAndWrite(t *testing.T,
...
-			maxReadInterval: maxReadInterval,
-			errCh:           errCh,
-			stopCh:          stopCh,
-			t:               t,
-		}
-		g.Go(reader.run)
-	}
+	// collect history records
+	var records historyRecords
+	var wg sync.WaitGroup
```
Comment (serathius) [inline concurrent_test.go:126]: having two sets of async goroutines synchronized on different primitives is hard to follow. Maybe move workers to sub function?

File: concurrent_test.go [hunk 2]
```diff
@@ -177,41 +199,53 @@ type readWorker struct {
...
 
 			if !reflect.DeepEqual(initialVal, val) {
 				return fmt.Errorf("read different values for the same key (%q), value1: %q, value2: %q",
 					selectedKey, formatBytes(initialVal), formatBytes(val))
 			}
 
+			clonedVal := make([]byte, len(val))
+			copy(clonedVal, val)
+
+			r.recordingCh <- historyRecord{
```
Comment (serathius) [inline concurrent_test.go:235]: In the robustness tests I wanted to avoid synchronizing on every operation. Reason is that slow recording can result in one fast worker starving other ones. For example reads being faster then writes result in writes being starved. 

I would recommend collecting local history in every worker separately and synchronizing only after all work being done.

File: concurrent_test.go [hunk 3]
```diff
@@ -321,3 +384,100 @@ func testResultsDirectory(t *testing.T) string {
...
+*/
+type workerType string
+
+const (
+	reader workerType = "reader"
+	writer workerType = "writer"
+)
+
+type historyRecord struct {
+	WorkerType workerType `json:"workerType,omitempty"`
```
Comment (serathius) [inline concurrent_test.go:401]: ClientId? Would recommend assigning unique id to each worker rather than grouping them by type. Types can change.

File: concurrent_test.go [hunk 4]
```diff
@@ -321,3 +384,100 @@ func testResultsDirectory(t *testing.T) string {
...
+}
+
+func (rs historyRecords) Less(i, j int) bool {
+	// Sorted by key firstly: all records with the same key are grouped together.
+	keyCmp := strings.Compare(rs[i].Key, rs[j].Key)
+	if keyCmp < 0 {
+		return true
+	} else if keyCmp > 0 {
+		return false
+	}
```
Comment (serathius) [inline concurrent_test.go:420]:
```suggestion
	if keyCmp != 0 {
		return keyCmp < 0
	}
```

File: concurrent_test.go [hunk 5]
```diff
@@ -321,3 +384,100 @@ func testResultsDirectory(t *testing.T) string {
...
+
+	// Sorted by txid
+	if rs[i].Txid < rs[j].Txid {
+		return true
+	} else if rs[i].Txid > rs[j].Txid {
+		return false
+	}
+
+	// Sorted by workerType: put writer before reader if they have the same txid.
+	if rs[i].WorkerType == writer {
```
Comment (serathius) [inline concurrent_test.go:430]: This is not worker type, but an operation type.

File: concurrent_test.go (hunk 3 above)
Comment (serathius) [inline concurrent_test.go:401]: This should be operationType.

File: concurrent_test.go [hunk 6]
```diff
@@ -321,3 +384,100 @@ func testResultsDirectory(t *testing.T) string {
...
+func collectHistoryRecords(recordingCh chan historyRecord) historyRecords {
+	var rs historyRecords
+	for record := range recordingCh {
+		rs = append(rs, record)
+	}
+	sort.Sort(rs)
+	return rs
+}
+
+func analyzeHistoryRecords(rs historyRecords) error {
```
Comment (serathius) [inline concurrent_test.go:450]: This is not a linearization validation as it doesn't account for operation execution time. It depends on Txid for order, making it very limited. For example it doesn't detect stale reads as View can just return old transation id, even though newer transaction was already committed. 

I'm not sure what concurrency issues this approach is mean to find or can find any. Please let me know if you need help with integrating linearizability checker https://github.com/anishathalye/porcupine like in robustness tests.

File: concurrent_test.go (hunk 5 above)
Comment (ahrtr) [reply concurrent_test.go:430]: updated

File: concurrent_test.go (hunk 4 above)
Comment (ahrtr) [reply concurrent_test.go:420]: thx & updated

File: concurrent_test.go (hunk 3 above)
Comment (ahrtr) [reply concurrent_test.go:401]: updated

File: concurrent_test.go (hunk 1 above)
Comment (ahrtr) [reply concurrent_test.go:126]: Makes sense, let me enhance it in a followup PR.

File: concurrent_test.go (hunk 2 above)
Comment (ahrtr) [reply concurrent_test.go:235]: Currently there is no starving issue, because each worker will sleep for a while between two transactions. But I agree we need improve it because I will support no sleep worker. Let me improve in followup PR.

File: concurrent_test.go (hunk 6 above)
Comment (ahrtr) [reply concurrent_test.go:446]: Depending on Txid is enough, because in bboltDB txid is strictly incremental. Any data written by a transaction should be visible to following reading transaction, otherwise it must be an issue. My understanding is that porcupine might generate false alarm, but in this test it will never generate false alarm unless it's a test bug. Also we are testing only local boltDB library instead of a distributed system, so I don't see any reason to use porcupine.

Note that there is stale read, a read transaction should always see the same data during the transaction lifecycle, and it's exactly what repeatable reading means.

**Comment by ahrtr** [comment]
Addressed some your comments, and I will resolve other comments in followup PR, thx. @serathius PTAL 

cc @ptabor as well.

File: concurrent_test.go (hunk 6 above)
Comment (ahrtr) [reply concurrent_test.go:446]: In summary, there are two points:
1. A read transaction should always see the same data view during its lifecycle.
2. Any data written by a writing transaction should be visible to any following reading transactions (with txid >= previous writing txid)

File: concurrent_test.go (hunk 1 above)
Comment (ahrtr) [reply concurrent_test.go]: done

File: concurrent_test.go (hunk 2 above)
Comment (ahrtr) [reply concurrent_test.go]: Done.

**Comment by ahrtr** [comment]
@ptabor @serathius PTAL, thx

File: concurrent_test.go (hunk 6 above)
Comment (ahrtr) [reply concurrent_test.go:477]: > You are mistaking serializability and linearizability.

Note that in bboltDB, only one writing transaction is allowed at a time. So in practice serializability is the same thing as linearizability.

Comment (serathius) [reply concurrent_test.go:477]: Goal of this tests is to drop all expectation and guesses, and let us validate them against the code behavior. Fact that it should not allow multiple writes, doesn't mean that we shouldn't test it.

Comment (ahrtr) [reply concurrent_test.go:477]: Note I am planning to add multiple writing transaction in followup PR.

Comment (ahrtr) [reply concurrent_test.go:477]: I can rename all `*linearizable*` to `*serializable*` when I start to support multiple writing transactions. thx for pointing this out.

Comment (ahrtr) [reply concurrent_test.go:477]: > I would recommend adding client side validation that transaction ID is always growing for each client, before we sort it.

Agreed. It's a good idea. Actually we also see such issue (txid isn't somehow strictly incremental). see https://github.com/etcd-io/bbolt/issues/402#issuecomment-1434220297

Let me do it in a separate PR. This PR is already too big.

File: concurrent_test.go [hunk 7]
```diff
@@ -106,34 +116,56 @@ func concurrentReadAndWrite(t *testing.T,
...
+	t.Log("Starting workers.")
+	records := runWorkers(t,
+		db, bucket, keys,
+		readerCount, minReadInterval, maxReadInterval,
+		minWriteInterval, maxWriteInterval, minWriteBytes, maxWriteBytes,
+		testDuration)
+
+	t.Log("Analyzing the history records.")
+	if err := analyzeHistoryRecords(records); err != nil {
+		t.Errorf("The history records are not linearizable:\n %v", err)
```
Comment (serathius) [inline concurrent_test.go:136]:
```suggestion
		t.Errorf("The history records are not serializable:\n %v", err)
```

File: concurrent_test.go [hunk 8]
```diff
@@ -106,34 +116,56 @@ func concurrentReadAndWrite(t *testing.T,
...
 
+	t.Log("Starting workers.")
+	records := runWorkers(t,
+		db, bucket, keys,
+		readerCount, minReadInterval, maxReadInterval,
+		minWriteInterval, maxWriteInterval, minWriteBytes, maxWriteBytes,
+		testDuration)
+
+	t.Log("Analyzing the history records.")
+	if err := analyzeHistoryRecords(records); err != nil {
```
Comment (serathius) [inline concurrent_test.go:135]:
```suggestion
	if err := validateSerializable(records); err != nil {
```

File: concurrent_test.go [hunk 9]
```diff
@@ -106,34 +116,56 @@ func concurrentReadAndWrite(t *testing.T,
...
+		t.Errorf("The history records are not linearizable:\n %v", err)
+	}
+
+	saveDataIfFailed(t, db, records)
```
Comment (serathius) [inline concurrent_test.go:139]: defer it to ensure it is executed even on fatal?

File: concurrent_test.go [hunk 10]
```diff
@@ -106,34 +116,56 @@ func concurrentReadAndWrite(t *testing.T,
...
 		return err
 	})
 	require.NoError(t, err)
 
+	t.Log("Starting workers.")
+	records := runWorkers(t,
+		db, bucket, keys,
+		readerCount, minReadInterval, maxReadInterval,
+		minWriteInterval, maxWriteInterval, minWriteBytes, maxWriteBytes,
+		testDuration)
```
Comment (serathius) [inline concurrent_test.go:132]: Can you create a configuration struct that will collect all the variables to ensure they are passed by name not by the order?

File: concurrent_test.go (hunk 7 above)
Comment (ahrtr) [reply concurrent_test.go]: done.

File: concurrent_test.go (hunk 8 above)
Comment (ahrtr) [reply concurrent_test.go]: done.

File: concurrent_test.go (hunk 9 above)
Comment (ahrtr) [reply concurrent_test.go:139]: If somehow panicking during test, then the records hasn't been collected yet. I may need change big change. Let me think about separately. thx

File: concurrent_test.go (hunk 10 above)
Comment (ahrtr) [reply concurrent_test.go:132]: agreed. Let me enhance this separately. thx

File: concurrent_test.go [hunk 11]
```diff
@@ -156,17 +216,13 @@ func concurrentReadAndWrite(t *testing.T,
...
 		t.Errorf("Received error: %v", err)
 	}
 
-	saveDataIfFailed(t, db)
-
-	// TODO (ahrtr):
-	//   1. intentionally inject a random failpoint.
-	//   2. validate the linearizablity: each reading transaction
-	//      should read the value written by previous writing transaction.
+	sort.Sort(rs)
```
Comment (serathius) [inline concurrent_test.go:224]: There is no reason that runWorkers should return ordered `historyRecords`. It's `validateSerializable` that expect history to be ordered. Please move sorting there to avoid implicit dependencies on order.

File: concurrent_test.go [hunk 12]
```diff
@@ -285,18 +365,41 @@ func formatBytes(val []byte) string {
 	return hex.EncodeToString(val)
 }
 
-func saveDataIfFailed(t *testing.T, db *btesting.DB) {
+/*
+*********************************************************
+Functions for persisting test data, including db file
+and operation history
+*********************************************************
+*/
```
Comment (serathius) [inline concurrent_test.go:373]: Please consider using standard godoc format https://go.dev/blog/godoc

Or better, don't comment and focus on function naming. This function has a good name and have only couple of logic lines. `saveDataIfFailed` is great name.

File: concurrent_test.go (hunk 11 above)
Comment (ahrtr) [reply concurrent_test.go]: Makes sense. Done.

File: concurrent_test.go [hunk 13]
```diff
@@ -321,3 +423,89 @@ func testResultsDirectory(t *testing.T) string {
 
 	return path
 }
+
+/*
+*********************************************************
+Data structures and functions for analyzing history records
```
Comment (serathius) [inline concurrent_test.go:429]: Introducing code section just means it should be a separate file/package.

File: concurrent_test.go (hunk 12 above)
Comment (ahrtr) [reply concurrent_test.go:372]: The comment isn't for a specific function or method, it's for a block of source code.  Note that I am planing to move them to separate files

File: concurrent_test.go (hunk 13 above)
Comment (ahrtr) [reply concurrent_test.go:429]: exactly. That's the plan. But in separate PR.

File: concurrent_test.go (hunk 6 above)
Comment (serathius) [reply concurrent_test.go]: Ups, I mistook [Serializable](https://jepsen.io/consistency/models/serializable) vs [Sequential](https://jepsen.io/consistency/models/sequential). Still validating `sequential` here would be useful and ok to be done in followup.

**Review by serathius — APPROVED** [review]
There are a lot of follow up, which is not to my liking. 

Still, as whole the PR is an improvement. Please make sure to do followups as postponing work is a bad example for other contributors. We usually ask other contributors to do all cleanups in the PR.

**Comment by ahrtr** [comment]
Thank you!

BoltDB test (to reproduce the data corruption issue) is my top priority. Some followup is my planned work.

File: concurrent_test.go (hunk 10 above)
Comment (ahrtr) [reply concurrent_test.go:132]: resolved in https://github.com/etcd-io/bbolt/pull/463

**Comment by ahrtr** [comment]
> do followups as postponing work is a bad example


Based on my previous experience, I am definitely not the blocker; instead, lack of active review is.
